core/tests/test-dependencies/root-files-mapping-generator.ts

```diff
@@ -0,0 +1,723 @@
+      });
+      this.referenceCache[dependency] = references;
+    }
+
+    return references.filter((reference: string) => {
+      if (ignoreModules) {
+        return !this.isFileAngularModule(reference);
+      }
+      return true;
+    });
+  }
+
+  /**
+   * Finds the root dependencies for the given file.
+   */
+  private getRootFilesOfFile(
+    file: string,
+    cache: Record<string, string[]> = {},
+    ignoreModules: boolean = true,
+    visited: Set<string> = new Set()
+  ): string[] {
+    if (cache[file]) {
+      return cache[file];
+    }
+    if (visited.has(file)) {
+      return [];
+    }
+    visited.add(file);
+
+    let references = this.getFilesWithDependency(file, ignoreModules);
+    if (references.length === 0 || this.pageModules.includes(file)) {
+      return [file];
+    }
+
+    const roots: string[] = [];
+    for (const reference of references) {
+      roots.push(
+        ...this.getRootFilesOfFile(reference, cache, ignoreModules, visited)
+      );
+    }
+
+    return Array.from(new Set(roots));
+  }
+
+  /**
+   * Gets the modules that correspond with the test suites.
+   */
+  private getTestSuiteModules(): string[] {
+    const testSuiteModules: string[] = [];
+    const testSuiteConfigFiles = fs
+      .readdirSync(CI_TEST_SUITE_CONFIGS_DIRECTORY)
+      .filter((file: string) => file.endsWith('.json'))
+      .map(file => path.join(CI_TEST_SUITE_CONFIGS_DIRECTORY, file));
+    for (const testSuiteConfig of testSuiteConfigFiles) {
+      const config = JSON.parse(fs.readFileSync(testSuiteConfig, 'utf-8'));
+      const suites = config.suites;
+      for (const suite of suites) {
+        testSuiteModules.push(suite.module);
+      }
+    }
+    return testSuiteModules;
+  }
+
+  /**
+   * Gets the valid root files.
+   */
+  private getValidRootFiles(): string[] {
+    const validRootFiles: string[] = [
+      ...this.pageModulesCoveredByRouteMapping,
+      ...this.getTestSuiteModules(),
+    ];
+    const rootFilesConfig = JSON.parse(
+      fs.readFileSync(ROOT_FILES_CONFIG_FILE, 'utf-8')
+    );
+    validRootFiles.push(...rootFilesConfig.VALID_ROOT_FILES);
+    validRootFiles.push(...rootFilesConfig.RUN_ALL_TESTS_ROOT_FILES);
+    return validRootFiles;
+  }
+
+  /**
+   * Validates the root files mapping.
+   */
+  private validateRootFilesMapping(
+    rootFilesMapping: Record<string, string[]>
+  ): void {
+    const rootFiles = Array.from(
+      new Set(Object.values(rootFilesMapping).flat())
+    );
+    const validRootFiles = this.getValidRootFiles();
+    const invalidRootFiles = rootFiles.filter((rootFile: string) => {
+      if (isFrontendTestFile(rootFile)) {
+        return false;
+      }
+      return !validRootFiles.includes(rootFile);
+    });
+    if (invalidRootFiles.length > 0) {
+      throw new Error(
+        'The following invalid root files were found when generating ' +
+          `the root files mapping:\n${invalidRootFiles.join('\n')}`
+      );
```

Comment:
In general, please try to add a piece of information about what to do with this error. Make it actionable.

Reply:
Ditto for other errors.

Reply:
Done.